File: Coding-Interviews/29-顺时针打印矩阵.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int         m;
    int         n;
    vector<int> spiralOrder(vector<vector<int>>& matrix)
    {
        m = matrix.size();
        n = matrix.empty() ? 0 : matrix[0].size();
        vector<int> ret;

        int border = 0;

        int M = m - 2 * border;
        int N = n - 2 * border;

        while (M > 0 && N > 0) {
            for (int i = 0; i < N; i++)
                ret.push_back(matrix[border][border + i]);
            for (int i = 0; i < M - 2; i++)
                ret.push_back(matrix[border + i + 1][border + N - 1]);
            if (M > 1) {
                for (int i = 0; i < N; i++)
                    ret.push_back(matrix[border + M - 1][border + N - 1 - i]);
            }
            if (N > 1) {
                for (int i = 0; i < M - 2; i++)
                    ret.push_back(matrix[border + M - 2 - i][border]);
            }
            border++;
            M = m - 2 * border;
            N = n - 2 * border;
        }
        return ret;
    }
};

File: Coding-Interviews/29-顺时针打印矩阵.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []

        def printCircle(startX, startY, lengthX, lengthY, res: List[int]):
            for i in range(0, lengthX):
                res.append(matrix[startX][startY+i])
            for i in range(0, lengthY-2):
                res.append(matrix[startX + 1 + i][startY+lengthX-1])
            if lengthY > 1:
                for i in range(0, lengthX):
                    res.append(matrix[startX + lengthY - 1]
                               [startY + lengthX - 1 - i])
            if lengthX > 1:
                for i in range(0, lengthY-2):
                    res.append(matrix[startX + lengthY - 2 - i][startY])
        res = []
        lenY = len(matrix)
        lenX = len(matrix[0])
        for i in range(0, (min(lenX, lenY) + 1)//2):
            printCircle(i, i, lenX - 2*i, lenY-2*i, res)
        return res


if __name__ == '__main__':
    print(Solution().spiralOrder([[7], [9], [6]]))
